Tests/kaas/k8s-node-distribution/k8s-node-distribution-check.py

```diff
@@ -0,0 +1,252 @@
+
+    return config
+
+
+def setup_logging(config_log):
+
+    logging.config.dictConfig(config_log)
+    loggers = [
+        logging.getLogger(name)
+        for name in logging.root.manager.loggerDict
+        if not logging.getLogger(name).level
+    ]
+
+    for log in loggers:
+        log.setLevel(config_log['level'])
+
+
+def initialize_config(config):
+    """Initialize the configuration for the test script"""
+
+    try:
+        with open(config.config_path, "r") as f:
+            config.logging = yaml.safe_load(f)['logging']
+    except OSError:
+        logger.warning(f"The config file under {config.config_path} couldn't be found, "
+                       f"falling back to the default config.")
+    finally:
+        # Setup logging if the config file with the relevant information could be loaded before
+        # Otherwise, we initialize logging with the included literal
+        setup_logging(config.logging or logging_config)
+
+    if config.kubeconfig is None:
+        raise ConfigException("A kubeconfig needs to be set in order to test a k8s cluster version.")
+
+    return config
+
+
+async def get_k8s_cluster_labelled_nodes(kubeconfig):
+    """Get the labels of the k8s cluster under test."""
+    _ = await kubernetes_asyncio.config.load_kube_config(kubeconfig)
+
+    nodes = list()
+    labels = [
+        "kubernetes.io/hostname",
+        "topology.kubernetes.io/zone",
+        "topology.kubernetes.io/region",
+        "node-role.kubernetes.io/control-plane"
+    ]
+
+    async with kubernetes_asyncio.client.ApiClient() as api:
+        core_api = kubernetes_asyncio.client.CoreV1Api(api)
+        result = await core_api.list_node()
+        for item in result.items:
+            node = dict()
+            node_labels = item.metadata.labels
+            for label in labels:
+                try:
+                    node[label] = node_labels[label]
+                except KeyError:
+                    # Node is not labelled as a control plane node
+                    pass
+            nodes.append(node)
+
+    return nodes
+
+
+def compare_labels(node_list, node_type="master"):
+
+    labels = {
+        "topology.kubernetes.io/region": set(),
+        "topology.kubernetes.io/zone": set(),
+        "kubernetes.io/hostname": set(),
+    }
+
+    for node in node_list:
+        for key in labels:
+            labels[key].add(node[key])
+
+    for label in labels:
+        if len(labels[label]) <= 1:
+            if len(labels[label]) == 1:
+                logger.warning(f"There seems to be no distribution across multiple {label.split('/')[1]}s "
+                               "or labels aren't set correctly across nodes.")
+            elif len(labels[label]) == 0:
+                logger.warning(f"The label for {label.split('/')[1]}s don't seem to be set for all nodes.")
+        else:
+            logger.info(f"The nodes are distributed across {str(len(labels[label]))} {label.split('/')[1]}s.")
+            return
```

Comment:
The `return` here comes as a little surprise to me. Are we happy if just one of the labels has more than 1 value? Maybe this can be documented a bit (preferably in the standard itself, but at least make a little remark here).

Reply:
Will document it.

Just to clarify it here as well, we're fine with one of these labels.
Basically, if we see that nodes are in different regions, it should (if the labeling is done correctly) not matter, if the (availability) zone or hostname is the same, since the same zone name or hostname could be used in different regions. Same thing goes for availability zones and hostnames.

If we don't want/can't rely on that, what would our way to go be? Throwing an error would probably invalidate quite a few clusters that are configured correctly (and some that are just labelled wrong, but that could always be the case).

I think being a bit more generous instead of too strict (also with regard to our time limit on this) is the better way.

Reply:
If I understand your answer correctly, we should check the labels in a certain order, right? In this case, region first, then zone, then hostname.

Reply:
This is not clear from the code. Yes, that is the order in which they are named when the dict `labels` is constructed, but you rely on an implementation detail, namely that this order is preserved when iterating over the dict. You have to document this assumption, or better yet, iterate over a tuple. At any rate, it should also be document why the order is important.

Reply:
I will document it in the text as well (and the tuple is coming)